### ISSUE TITLE
Targeting Beam as Training Target Causes Dimension Error

### ISSUE THREAD

**Issue by npwynands** [rwth-i6/returnn#304]
This issue is related to the following StackOverflow question: [Teacher-Student System: Training Student With k Target Sequences for Each Input Sequence](https://stackoverflow.com/questions/62344553/teacher-student-system-training-student-with-k-target-sequences-for-each-input/62359741#62359741)

I want to configure a teacher-student system, where a teacher seq2seq model generates a top-*k* list of hypotheses, which are used to train a student seq2seq model.

I select the top-*k* hypotheses list from the teacher’s `ChoiceLayer` (or output layer) by:
```python
"teacher_hypotheses": {
    "class": "copy", "from": ["extra.search:teacher_output"],
    "register_as_extern_data": "teacher_hypotheses_data"
}
```
The output `Data` of that layer has a batch axis length `batch_size`=*k*=4 times the length of the input `Data`’s batch axis length (cf. doc and code of: `Data.copy_extend_with_beam`, `SearchChoices.translate_to_common_search_beam`).

`teacher_hypotheses_data` is selected as the student’s training target. But this leads to the following error:
```shell
TensorFlow exception: assertion failed: [shape[0]:] [92] [!=] [dim:] [23]
     [[node student_output/rec/subnet_base/check_seq_len_batch_size/check_input_dim/assert_equal_1/Assert/Assert (defined at home/philipp/Documents/bachelor-thesis/returnn/returnn-venv/lib/python3.7/site-packages/tensorflow_core/python/framework/ops.py:1748) ]]
```
Which is, I assume, due to the issue that the target data of the student, the hypotheses list, has a batch axis length *k*=4 times longer than the one of the student’s input data/encoder state data.

According to Albert’s comment, this might be a bug. I attached you a simplified version of my config, which causes the same error. In this case it uses dummy data, so no datasets are needed.

**Environment**:
- OS: Ubuntu 18 (local machine, not the cluster)
- Python 3.7.5
- TensorFlow 1.15.2
- RETURNN (commit 8595f10b03067b0a2253d0ee65d7dd7efd62ce76; 27th May, 15:40)

**Example Config**:
[beam-as-target.py.txt](https://github.com/rwth-i6/returnn/files/4779979/beam-as-target.py.txt)

**Comment by albertz** [comment]
This should be fixed now (via PR #310). Please test.

**Comment by albertz** [comment]
Btw, if you want to see how to work on a fix like this:

* I first created `test_target_with_beam` (following your example, possibly simplifying it a bit, also reducing dimensions) which recreates the issue.
* Then understand the problem itself. The main problem here was that `TFUtil.SearchBeam` and `TFNetworkRecLayer.SearchChoices` was somewhat orthogonal to each other, and via `register_as_extern_data`, the search choices information was lost.
* Then come up with a solution. In this case, this was not so simple. The easiest solution I came up (for now) is a new `_map_search_beam_to_search_choices`.
* Then see if that fixes the problem, and if not, repeat in understanding the (maybe next / other / different) problem, and again come up with further solutions. In this example, there were about 15 different commits needed. I would say this is probably difficult if you don't know the internals of RETURNN already to some degree. But it would also be a good way to learn more about the RETURNN internals. Although it is maybe difficult to judge about the best solutions. In any case, the solution should be generic and clean, not hacky. It should improve the overall code quality.

So, if you are curious, check `test_target_with_beam`, and check all the recent commits of this PR.
